### ISSUE TITLE
Datetime issue

### ISSUE THREAD

**Issue by alexsantm2019** [briannesbitt/Carbon#1713]
Hi Everyone,
Yesterdary I was working with mi app in Laravel 5.8, and suddenly my APIS did not work. The error mentionated in console was: 

message | DateTime::setTimestamp() expects parameter 1 to be int, float given
-- | --
exception | ErrorException
file | C:\xampp\htdocs\atc-core\vendor\nesbot\carbon\src\Carbon\Traits\Units.php
line | 109

I use laravel as backend, and I work with created_at, updated_at and deleted_at fields. 

Is there an issue or I have to set a extra config in my files??

thanks in advanced

**Comment by kylekatarnls** [comment]
Some informations we ask in our issue template are missing, to get support please ensure to give at least:

- [ ] your Carbon version;
- [ ] your PHP version;
- [ ] a minimal code chunk to reproduce your case;
- [x] the current output of this code;
- [x] the output you expect.

Thanks.

**Comment by alexsantm2019** [comment]
Thanks:
In composer.json sayd: "nesbot/carbon": "^1.26.3 || ^2.0",
PHP Versión is 7.1
The current ouput in console is (when I execute an API):

message | DateTime::setTimestamp() expects parameter 1 to be int, float given
-- | --
exception | ErrorException
file | C:\xampp\htdocs\atc-core\vendor\nesbot\carbon\src\Carbon\Traits\Units.php
line | 109

The outpu I expect is a simple query from my database. 
Thanks

**Comment by kylekatarnls** [comment]
> In composer.json sayd: "nesbot/carbon": "^1.26.3 || ^2.0",

It's not your version, there is a lot of version from 1.26.3 to the latest ^2, as the issue template explains: Run the command `composer show nesbot/carbon`

I understood expected/current output, that's why I checked those checkboxes. Still I miss:

- [ ] a minimal code chunk to reproduce your case;

**Comment by alexsantm2019** [comment]
This is my nesbot/carbon version:

name     : nesbot/carbon
descrip. : A simple API extension for DateTime.
keywords : date, datetime, time
versions : * 2.17.1
type     : library
license  : MIT License (MIT) (OSI approved) https://spdx.org/licenses/MIT.html#licenseText
source   : [git] https://github.com/briannesbitt/Carbon.git 96acbc0c03782e8115156dd4dd8b736267155066
dist     : [zip] https://api.github.com/repos/briannesbitt/Carbon/zipball/96acbc0c03782e8115156dd4dd8b736267155066 96acbc0c03782e8115156dd4dd8b736267155066
path     : C:\xampp\htdocs\atc-core\vendor\nesbot\carbon
names    : nesbot/carbon

autoload
psr-4
Carbon\ => src/Carbon/

requires
ext-json *
php ^7.1.8 || ^8.0
symfony/translation ^3.4 || ^4.0

requires (dev)
friendsofphp/php-cs-fixer ^2.14 || ^3.0
kylekatarnls/multi-tester ^1.1
phpmd/phpmd ^2.6
phpstan/phpstan ^0.11
phpunit/phpunit ^7.5 || ^8.0
squizlabs/php_codesniffer ^3.4


A minimal code Chunk:
1. In model:

<?php

namespace App\Core\Eloquent\Frontend\Producto;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;
use Carbon;

class Localidad extends Model
{
    protected $connection = 'mysql2';
    protected $table = 'localidad';
    protected $primaryKey = 'loc_id';
    use SoftDeletes;
    
    //protected $dateFormat = 'Y-m-d H:i:s';
    protected $dateFormat = 'Y-m-d H:i:s';
    //protected $dateFormat = 'U';

    protected $dates = ['created_at', 'updated_at','deleted_at'];
    protected $fillable=[
        'loc_id',
        'loc_region',
        'loc_nombre',
        'loc_codigo',
        'loc_tipo',
        'loc_padre_pais',
        'loc_padre_provincia',
        'loc_padre_ciudad',
    ];
}

2. My API:
    public function getCity()
    {
        $localidad = Localidad::where('loc_padre_provincia', '=', NULL)->get();
        if($localidad){return response()->json(['data' =>$localidad], 200);}
        else{ return response()->json(['message' =>"No existen Registros"], 404);
        }
    }

**Comment by kylekatarnls** [comment]
Hi, thanks. Line 109 of Units.php in Carbon 2.17.1 is:

```php
return $this->setTimestamp($this->getTimestamp() + $value);
```

To confirm it you can change this line to:
```php
return $this->setTimestamp((int) $this->getTimestamp() + $value);
```

If it fixes the problem, so it's the right entry-point.

So now I can know what is calling setTimestamp, but the code chunk does not help me to reproduce the bug since I don't know what is calling addRealUnit() (should be methods like addSeconds/addMinutes/etc.), if it's not in your own code, it's in Laravel and so I need the complete stack trace of the error to find the code in the Laravel source code.

Note than typing in PHP from float to int should be soft. For example if I do:
```php
Carbon::now()->setTimestamp(0.5);
```

0.5 is converted silently to 0 and no error is thrown.

Do you use `declare(strict_types=1);`?

**Comment by kylekatarnls** [comment]
After a deeper check, DateTime::setTimestamp() does not have strong typing neither in PHP 7.1 nor 7.2 and CarbonInterface is compatible. It means even `strict_types=1` should not throw any error for this specific typing even if float is passed.

Did you extend Carbon in any way that could override setTimestamp and add this strong typing?

Please search `setTimestamp(int` in your source code including the `vendor` directory. If it's in your code, you should remove the `int` typing and it should work just fine again. If it's a package, please give me the name and version of this package and the file and line where there is the typing.

Note that nothing changed about this neither in PHP nor in Carbon. So a change in your source or an update in a package could have caused the error, but it's probably not related to Carbon itself.

**Comment by kylekatarnls** [comment]
I'm still not sure how it could have changed but as a safety-guard, we will add explicit typing casts to methods called by Laravel to avoid this to happen.

**Comment by bassbiteye45** [comment]
i have the same issue could you help me pls ?

**Comment by kylekatarnls** [comment]
This issue has been fixed in Carbon 2.18.0. If you have an earlier version, it can't be the same issue, so open a new one with the required details. If you have a older version, upgrade your version of Carbon.

**Comment by bassbiteye45** [comment]
thank you
